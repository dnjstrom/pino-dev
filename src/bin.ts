#!/usr/bin/env node

import fs from "fs";
import pump from "pump";
import split from "split2";
import { Transform } from "readable-stream";
import { prettifierFactory } from "./index";

const prettify = prettifierFactory();
const prettyTransport = new Transform({
  objectMode: true,
  transform(chunk, enc, cb) {
    const line = prettify(chunk.toString());
    if (line === undefined) {
      return cb();
    }
    cb(undefined, line);
  },
});

const join = new Transform({
  objectMode: true,
  transform(
    chunk: any,
    encoding: BufferEncoding,
    callback: (error?: Error, data?: any) => void
  ) {
    callback(undefined, chunk.toString() + "\n");
  },
});

pump(process.stdin, split(), prettyTransport, join, process.stdout);

// https://github.com/pinojs/pino/pull/358
if (!process.stdin.isTTY && !fs.fstatSync(process.stdin.fd).isFile()) {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  process.once("SIGINT", () => {});
}
