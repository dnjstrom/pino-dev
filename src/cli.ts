#!/usr/bin/env node

import fs from "fs";
import pump from "pump";
import split from "split2";
import { Transform } from "readable-stream";
import { prettifierFactory } from "./index";
import args from "args";
import bourne from "@hapi/bourne";
import { debug } from "./debug";

args.option(
  ["m", "map"],
  "Map arbitrary incoming properties to semantic pino-dev properties using json."
);

const opts = args.parse(process.argv);
debug(`pino-dev started with options: ${JSON.stringify(opts, null, 2)}`);

const propertyMap = opts.map
  ? bourne.parse(opts.map, { protoAction: "remove" })
  : undefined;

const prettify = prettifierFactory({ propertyMap });
const prettyTransport = new Transform({
  objectMode: true,
  transform(chunk, enc, cb) {
    const line = prettify(chunk.toString());
    if (line === undefined) {
      return cb();
    }
    cb(undefined, line);
  },
});

pump(process.stdin, split(), prettyTransport, process.stdout);

// https://github.com/pinojs/pino/pull/358
if (!process.stdin.isTTY && !fs.fstatSync(process.stdin.fd).isFile()) {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  process.once("SIGINT", () => {});
}
