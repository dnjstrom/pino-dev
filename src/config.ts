import JoyCon from "joycon";
import bourne from "@hapi/bourne";
import { debug } from "./debug";
import { Config } from "./types";
import { PartialDeep } from "type-fest";

const joycon = new JoyCon({
  files: ["package.json", "pino-dev.config.json", "pino-dev.config.js"],
  packageKey: "pino-dev",
  parseJSON: (str) => bourne.parse(str, { protoAction: "remove" }),
});

const { path, data } = joycon.loadSync();
const configFileData: PartialDeep<Config> = data ?? {};

if (path) {
  debug(`Using config from ${path}.`);
}

export const DEFAULT_CONFIG: Config = {
  newline: "\n",
  timeFormat: "HH:mm:ss.SSS",
  propertyMap: {
    msg: "msg",
    level: "level",
    ns: "ns",
    name: "name",
    stack: "stack",
    time: "time",
    "req.method": "req.method",
    "req.url": "req.url",
    "res.statusCode": "res.statusCode",
    responseTime: "responseTime",
  },
};

export const mergeConfig = (
  base: Config,
  ...additionalConfigs: Array<PartialDeep<Config>>
): Config =>
  additionalConfigs.reduce<Config>((agg, config) => {
    return {
      ...agg,
      ...config,
      propertyMap: {
        ...agg.propertyMap,
        ...config.propertyMap,
      },
    };
  }, base);

export const config = mergeConfig(DEFAULT_CONFIG, configFileData);
