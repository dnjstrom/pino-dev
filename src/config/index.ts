import JoyCon from "joycon";
import { parse } from "@hapi/bourne";
import { debug } from "../debug";
import { PartialDeep } from "type-fest";
import { DEFAULT_CONFIG } from "./DEFAULT_CONFIG";
import { mergeConfig } from "./mergeConfig";
import { PropertyMap } from "../parse/PropertyMap";

export type Config = {
  newline: string;
  timeFormat: string;
  propertyMap: PropertyMap;
  colorize?: boolean | "true" | "false";
};

const joycon = new JoyCon({
  files: ["package.json", "pino-dev.config.json", "pino-dev.config.js"],
  packageKey: "pino-dev",
  parseJSON: (str) => parse(str, { protoAction: "remove" }),
});

const { path, data } = joycon.loadSync();
const configFileData: PartialDeep<Config> = data ?? {};

if (path) {
  debug(`Using config from ${path}.`);
}

export const config = mergeConfig(DEFAULT_CONFIG, configFileData);
