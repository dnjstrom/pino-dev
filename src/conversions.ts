import chalk, { Chalk } from "chalk";

export const levelToString = (level: number): string => {
  switch (level) {
    case 60:
      return "Fatal";
    case 50:
      return "Error";
    case 40:
      return "Warn";
    case 30:
      return "Info";
    case 20:
      return "Debug";
    case 10:
      return "Trace";
    default:
      return "Unknown";
  }
};

export const levelToColor = (level: number): Chalk => {
  switch (level) {
    case 60:
      return chalk.red.bold;
    case 50:
      return chalk.red;
    case 40:
      return chalk.yellow;
    case 30:
      return chalk.green;
    case 20:
      return chalk.blue;
    case 10:
      return chalk.cyan;
    default:
      return chalk.reset;
  }
};

export const statusCodeToColor = (statusCode: number): Chalk => {
  if (statusCode >= 500 && statusCode <= 599) {
    return chalk.red;
  } else if (statusCode >= 400 && statusCode <= 499) {
    return chalk.yellow;
  } else if (statusCode >= 300 && statusCode <= 399) {
    return chalk.blue;
  } else if (statusCode >= 200 && statusCode <= 299) {
    return chalk.green;
  } else {
    return chalk.magenta;
  }
};
