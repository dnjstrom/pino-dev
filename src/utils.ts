const repeatString = (amount: number, str: string) =>
  new Array(amount + 1).join(" ");

export const leftpad = (amount: number, str: string, padCharacter = " ") =>
  repeatString(amount, padCharacter) + str;

export const padLinesWithSpaces = (
  amount: number,
  str: string,
  newline: string
) => {
  const spaces = repeatString(amount, " ");
  return str.replace(new RegExp(`${newline}`, "g"), `${newline}${spaces}`);
};

export const getDeep = (keys: string[], obj: any): any => {
  const [key, ...remainingKeys] = keys;

  if (obj == null) {
    return undefined;
  } else if (key == null) {
    return obj;
  }

  return getDeep(remainingKeys, obj[key]);
};

export const setDeep = (obj: any = {}, keys: string[], value: any): any => {
  const [key, ...remainingKeys] = keys;

  if (key == null) {
    return obj;
  } else if (remainingKeys.length === 0) {
    obj[key] = value;
    return obj;
  }

  obj[key] = setDeep(obj[key] ? obj[key] : {}, remainingKeys, value);

  return obj;
};

export const joinWith = (separator: string, items: unknown[]) =>
  items.filter(Boolean).join(separator);

export const words = (...args: unknown[]) => joinWith(" ", args);
