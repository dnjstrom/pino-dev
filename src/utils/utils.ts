import { Falsy } from "../types";

const repeatString = (amount: number, str: string): string =>
  new Array(Math.max(amount, 0) + 1).join(str);

export const leftPad = (
  amount: number,
  str: string,
  padCharacter = " "
): string => repeatString(amount, padCharacter) + str;

export const padLinesWithSpaces = (
  amount: number,
  str: string,
  newline: string
): string => {
  const spaces = repeatString(amount, " ");
  return str.replace(new RegExp(`${newline}`, "g"), `${newline}${spaces}`);
};

export const getDeep = <T = unknown>(
  keys: string[],
  obj: unknown
): T | undefined => {
  const [key, ...remainingKeys] = keys;

  if (obj == null) {
    return undefined;
  } else if (key == null) {
    return obj as T;
  }

  return getDeep(remainingKeys, (obj as Record<string, unknown>)[key]);
};

export const setDeep = <T = unknown>(
  obj: Record<string, unknown>,
  keys: string[],
  value: T
): Record<string, unknown> => {
  const [key, ...remainingKeys] = keys;

  if (key == null) {
    throw new Error(
      `Invalid path "${keys}" when setting "${value}" in \`${JSON.stringify(
        obj
      )}\`.`
    );
  } else if (remainingKeys.length === 0) {
    obj[key] = value;
    return obj;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const parent = obj as any;
  const child = parent[key] ? parent[key] : {};
  obj[key] = setDeep(child, remainingKeys, value);

  return obj;
};

/**
 * Concatenates parameters with separator, omitting any falsy values.
 */
export const joinWith = (
  separator: string,
  items: Array<string | Falsy>
): string => items.filter(Boolean).join(separator);
