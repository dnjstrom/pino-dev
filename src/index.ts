import bourne from "@hapi/bourne";
import { debug } from "./debug";
import { getDeep, setDeep } from "./utils";
import { format } from "./format";
import { config, mergeConfig } from "./config";
import { Config, Input, PropertyMap } from "./types";
import { PartialDeep } from "type-fest";

const mapProperties = (
  propertyMap: PropertyMap,
  input: Record<string, unknown>
): Record<string, unknown> =>
  Object.entries(propertyMap).reduce((agg, [to, from]) => {
    if (typeof from === "boolean" && !from) {
      return agg;
    } else if (typeof from !== "string") {
      throw new Error(
        `Invalid property mapping for "${to}": ${JSON.stringify(from)}.`
      );
    }

    const value = getDeep(from.split("."), input);

    if (value !== undefined) {
      setDeep(agg, to.split("."), value);
    }

    return agg;
  }, {});

const validateInput = (
  propertyMap: PropertyMap,
  input: Record<string, unknown>
): Input => {
  const mapped = mapProperties(propertyMap, input);

  if (mapped.msg == null) {
    throw new Error("Input is missing `msg`-property");
  }

  return mapped as Input;
};

export const prettifierFactory = (
  options: PartialDeep<Config> = {}
): ((line: unknown) => string) => {
  const opts = mergeConfig(config, options);

  debug(`Using config ${JSON.stringify(opts, null, 2)}.`);

  return (line: unknown) => {
    let input: Input;

    try {
      const parsed =
        typeof line === "string"
          ? bourne.parse(line, { protoAction: "remove" })
          : line;

      input = validateInput(opts.propertyMap, parsed);
    } catch (err) {
      debug(`Error parsing input \`${line}\`.`);
      debug(err);
      return line + config.newline;
    }

    return format(input, opts) + config.newline;
  };
};
