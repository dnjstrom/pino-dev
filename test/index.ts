process.env.FORCE_COLOR = "0";

import { prettifierFactory } from "../src";
import { examples } from "./utils/examples";

describe("formatting examples", () => {
  let prettifier: ReturnType<typeof prettifierFactory>;

  beforeEach(() => {
    prettifier = prettifierFactory();
  });

  examples.forEach((input, index) => {
    it(`formats example #${index} correctly`, () => {
      expect(prettifier(JSON.stringify(input))).toMatchSnapshot();
    });
  });
});
