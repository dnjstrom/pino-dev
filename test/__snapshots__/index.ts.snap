// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatting examples formats example #0 correctly 1`] = `
"\\"miscellaneous text\\"
"
`;

exports[`formatting examples formats example #1 correctly 1`] = `
"02:56:15 Trace: An info message
"
`;

exports[`formatting examples formats example #2 correctly 1`] = `
"02:56:15 Debug: A debug message
"
`;

exports[`formatting examples formats example #3 correctly 1`] = `
"02:56:15 Info: An info message
"
`;

exports[`formatting examples formats example #4 correctly 1`] = `
"02:56:15 Warn: A warning message
"
`;

exports[`formatting examples formats example #5 correctly 1`] = `
"02:56:15 Error: An error message
"
`;

exports[`formatting examples formats example #6 correctly 1`] = `
"02:56:15 Fatal: A fatal message
"
`;

exports[`formatting examples formats example #7 correctly 1`] = `
"02:56:15 Info: 	    A message with leading whitespace
"
`;

exports[`formatting examples formats example #8 correctly 1`] = `
"02:56:15 Info: A message with ending whitespace    	
"
`;

exports[`formatting examples formats example #9 correctly 1`] = `
"02:56:15 Info: A multi-line message. Lorem ipsum
               dolor sit amet, consectetur adipiscing elit. Vivamus vitae orci volutpat,
               finibus augue vitae, pellentesque nibh. Maecenas lacus erat, maximus sit amet ligula sed,
               porttitor tincidunt felis. Donec dapibus eget est eu fermentum. Sed eu augue turpis.
               Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae;
               Vivamus vehicula ante nisi, id porta odio blandit ut.
"
`;

exports[`formatting examples formats example #10 correctly 1`] = `
"02:56:15 Info: A message with an object parameter
"
`;

exports[`formatting examples formats example #11 correctly 1`] = `
"02:56:15 Error: An error happened somehow
    Error: Oh noes!
    at /some/path/to/file:23:98
    at step (/some/path/to/file:23:98)
    at Object.next (/some/path/to/file:23:98)
    at /some/path/to/file:23:98
    at new Promise (<anonymous>)
    at __awaiter (/some/path/to/file:23:98)
    at generateExamples (/some/path/to/file:23:98)
    at Object.<anonymous> (/some/path/to/file:23:98)
    at Module._compile (internal/some/path/to/file:23:98)
    at Module.m._compile (/some/path/to/file:23:98)
"
`;

exports[`formatting examples formats example #12 correctly 1`] = `
"02:56:15 Info: Meta data via child logger
"
`;

exports[`formatting examples formats example #13 correctly 1`] = `
"02:56:15 [My Logger] Info: A logger with a name set
"
`;

exports[`formatting examples formats example #14 correctly 1`] = `
"02:56:15 Info: A logger without the base child logger
"
`;

exports[`formatting examples formats example #15 correctly 1`] = `
"\\"\\"
"
`;
